rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala

```diff
@@ -0,0 +1,319 @@
+              case None => Left(MethodNotFound(s"Could not find reporting values '${method.classParameter.value}' method '${methodId.value}' in Technique '${technique.bundleName.value}'",None))
+              case Some(value) => Right (<VALUE>{value}</VALUE>)
+            }
+          } yield {
+            <SECTION component="true" multivalued="true" name={methodId.value}>
+              <REPORTKEYS>
+                {spec}
+              </REPORTKEYS>
+            </SECTION>
+          }
+      }
+    }
+    // Regroup method calls from which we expect a reporting
+    // We filter those starting by _, which are internal methods
+    val expectedReportingMethodsWithValues =
+      for {
+        (name, methodCalls) <- technique.methodCalls.groupBy(_.methodId).toList.sortBy(_._1.value)
+        if ! name.value.startsWith("_")
+      } yield {
+        (name, methodCalls)
+      }
+
+     for {
+       reportingSection     <- sequence(expectedReportingMethodsWithValues)((reportingValuePerMethod _).tupled)
+       agentSpecificSection <- sequence(agentSpecific)(_.agentMetadata(technique, methods))
+     } yield {
+       <TECHNIQUE name={technique.name}>
+         <DESCRIPTION>technique.description</DESCRIPTION>
+         {agentSpecificSection}
+         <SECTIONS>
+           {reportingSection}
+         </SECTIONS>
+       </TECHNIQUE>
+     }
+  }
+
+  // Write and commit all techniques files
+  def writeAll(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, committer : EventActor) : Result[Seq[String]] = {
+    for {
+      agentFiles <- writeAgentFiles(technique, methods, modId, committer)
+      res <- writeMetadata(technique, methods, modId, committer)
+    } yield {
+      res +: agentFiles
+    }
+  }
+
+  def writeAgentFiles(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, commiter : EventActor) : Result[Seq[String]] = {
+    for {
+
+      // Create/update agent files, filter None
+      files  <- sequence(agentSpecific)(_.writeAgentFile(technique, methods)).map(_.flatten)
+      _ <- sequence(files)(archiver.commitFile(technique, _, modId, commiter , s"Commiting Technique '${technique.bundleName.value}' file for agent " ))
+    } yield {
+      files
+    }
+  }
+
+  def writeMetadata(technique : Technique, methods: Map[BundleName, GenericMethod], modId : ModificationId, commiter : EventActor) : Result[String] = {
+
+    val metadataPath = s"techniques/ncf_techniques/${technique.bundleName.value}/${technique.version.value}/metadata.xml"
+
+    val path = s"/var/rudder/configuration-repository/${metadataPath}"
+    for {
+      content <- techniqueMetadataContent(technique, methods)
+      _ <- execute{
+             val filePath = Paths.get(path)
+             if (!Files.exists(filePath))
+               Files.createFile(filePath)
+             scala.xml.XML.save(path, content)
+           } {case e => IOError("",Some(e))}
```

Comment:
never ever an empty message. It is what we will always give to the user, and need to explain the "business" pb.

Reply:
Just one i missed :(